chainerx_cc/chainerx/routines/math.cc

```diff
@@ -525,6 +525,77 @@ Array IfGreaterElse(const Array& x1, Scalar x2, Scalar pos, const Array& neg) {
 
 }  // namespace
 
+namespace {
+
+Array IfGreaterElse(const Array& x1, const Array& x2, const Array& pos, const Array& neg) {
+    CheckEqual(x1.dtype(), x2.dtype());
+    CheckEqual(x1.shape(), x2.shape());
+    Array out = EmptyLike(x1, x1.device());
+    {
+        NoBackpropModeScope scope{};
+        x1.device().IfGreaterElseAAAA(x1, x2, pos, neg, out);
+    }
+
+    {
+        BackwardBuilder bb{"if_greater_else", {pos, neg}, out};
```

Comment:
This backward implementation also duplicates with `IfGreaterElseImpl`. How about the following implementation?
```
CheckEqual(x1.dtype(), x2.dtype());
CheckEqual(x1.shape(), x2.shape());
Array out = EmptyLike(x1);
IfGreaterElseImpl(x1, x2, pos, neg, out);
return out;
```

Reply:
Defining `Array out` within `MinimumImpl` shadows the `out` parameter. Also, the return statement will require `Array` return type instead of `void`. So, would an implementation like this be sufficient?
`
CheckEqual(x1.dtype(), x2.dtype());
CheckEqual(x1.shape(), x2.shape());
IfGreaterElseImpl(x1, x2, pos, neg, out);
`

Reply:
The current implementation of `MinimumImpl` seems to be OK, but both `IfGreaterElse` and `IfGreaterElseImpl` have backward implementations.

Reply:
@asi1024 Sorry, I probably misunderstood your previous comment. I will make the required modifications to `IfGreaterElseImpl`.

Reply:
> This backward implementation also duplicates with `IfGreaterElseImpl`. How about the following implementation?
> 
> ```
> CheckEqual(x1.dtype(), x2.dtype());
> CheckEqual(x1.shape(), x2.shape());
> Array out = EmptyLike(x1);
> IfGreaterElseImpl(x1, x2, pos, neg, out);
> return out;
> ```

As I understand it, `IfGreaterElseImpl` has to make use of `IfGreaterElse` in its backward. So, wouldn't replicating the backward implementations in `IfGreaterElse` and `IfGreaterElseImpl` be the only possible option apart from removing `IfGreaterElse` completely and only having `IfGreaterElseImpl` ? Sorry if I'm missing out on something.

Reply:
Ah, now I understand it. How about retaining a bitmask `x1 > x2` instead of raw data of `x1` and `x2`? It not only resolve the problem in my comment but also reduce memory consumption. The implementation will be as follows:
```
const Array& gout = *bctx.output_grad();
bctx.input_grad() = gout * mask.AsType(gout.dtype());  // or (LogicalNot(mask).Astype(...))
```

Reply:
Great. Thanks for the idea @asi1024 . Will make the required changes.

Reply:
@asi1024 Sorry for the delayed reply. Do you suggest retention of the bitmask in `IfGreaterElse` only and not `IfGreaterElseImpl` to avoid duplication of their backwards?

Reply:
![mask_impl](https://user-images.githubusercontent.com/42710132/55273773-b04fdd80-52f5-11e9-9dea-d594476a4cab.png)
@asi1024 This is my attempt at an implementation of your idea. Please let me know if any modifications can be made? Note: I've left `IfGreaterElseImpl` unchanged. (Still possesses the normal backward implementation).

Reply:
If you rewrite `IfGreaterElseImpl` in the same way, `IfGreaterElse` seems to be able to be removed.

Reply:
`mask` and `not_mask` should be of `bool` dtype. How about the following implementation?
```
Array mask = Greater(x1, x2);
Array not_mask = LogicalNot(mask);
```

Reply:
A while ago, some PRs about mixed-dtypes are merged in the master branch. Could you merge master branch and support mixed-dtypes in `chainerx.minimum`?

Reply:
I have made the required changes to the masks. I have also rebased the branch. Hope that is okay.